### PR TITLE
Multihead matmul sparse

### PR DESCRIPTION
**Pull request by minghaoBD**
### PR types
Performance optimization

### PR changes
OPs

### Describe
Made several optimizations in this PR:
1. Reduce the runtime memory consumptions by introducing shared_ptr in spmm_plugin.cu.
2. Support FP16 inference by changing the bias from FP32 to FP16 while in fp16-mode inference.
3. Apply the spmm_plugin into the fused multihead_matmul op to further enhance the inference performance.
4. Add a UT accordingly.

### CODE REVIEW

File: paddle/fluid/inference/tensorrt/CMakeLists.txt [hunk 1]
```diff
@@ -6,6 +6,6 @@ nv_library(tensorrt_engine SRCS engine.cc trt_int8_calibrator.cc DEPS ${GLOB_OPE
 endif()
 nv_library(tensorrt_op_teller SRCS op_teller.cc DEPS framework_proto device_context boost)
 nv_test(test_tensorrt SRCS test_tensorrt.cc DEPS dynload_cuda device_context dynamic_loader)
-nv_test(test_tensorrt_engine SRCS test_engine.cc DEPS dynload_cuda tensorrt_engine)
+nv_test(test_tensorrt_engine SRCS test_engine.cc test_dynamic_engine DEPS dynload_cuda tensorrt_engine tensorrt_plugin)
```
Comment (b3602sss) [inline paddle/fluid/inference/tensorrt/CMakeLists.txt]: 少了.cc?

File: paddle/fluid/framework/ir/CMakeLists.txt [hunk 2]
```diff
@@ -101,6 +101,7 @@ pass_library(matmul_scale_fuse_pass inference)
 pass_library(gpu_cpu_map_matmul_to_mul_pass inference)
 pass_library(mixed_precision_configure_pass inference)
 pass_library(replace_dense_with_sparse_pass inference)
+pass_library(replace_dense_multihead_matmul_with_sparse_pass inference)
```
Comment (b3602sss) [inline paddle/fluid/framework/ir/CMakeLists.txt]: 把replace_dense_with_sparse_pass改成replace_dense_fc_with_sparse_pass或者把replace_dense_multihead_matmul_with_sparse_pass合入到replace_dense_with_sparse_pass。

File: paddle/fluid/inference/tensorrt/CMakeLists.txt (hunk 1 above)
Comment (minghaoBD) [reply paddle/fluid/inference/tensorrt/CMakeLists.txt]: Done, thanks

File: paddle/fluid/framework/ir/CMakeLists.txt (hunk 2 above)
Comment (minghaoBD) [reply paddle/fluid/framework/ir/CMakeLists.txt]: replace_dense_with_sparse_pass -> replace_dense_fc_with_sparse_pass

**Pull request closed without merging**